File: bzz-spark/src/main/scala/com/spark/scala/study/SortWord.scala
package com.spark.scala.study

import org.apache.spark.{SparkConf, SparkContext}

/**
  * @Title: SortWord
  * @ProjectName bzz-cloud
  * @Auther: clouder
  * @Date: 2018-07-14-0:59
  * @Description:
  */
object SortWord {

  def main(args: Array[String]): Unit = {
    val sparkConf = new SparkConf().setAppName("SortWord").setMaster("local")
    val sparkContext = new SparkContext(sparkConf)

    val sparkRdd = sparkContext.textFile("D:\\app\\word.txt")

    val wordRdd = sparkRdd.flatMap(word => { word.split(" ")})
    val wordCountRdd = wordRdd.map(word => (word,1))
    val wordReduceRdd = wordCountRdd.reduceByKey(_ + _)
    wordReduceRdd.foreach(word =>{
      println(word._1 + ":" +word._2)
    })
    val converRDD = wordReduceRdd.map(word => (word._2,word._1))
    val secondRdd = converRDD.sortByKey(false)
    val resultRdd = secondRdd.map(word => (word._2,word._1))
    resultRdd.foreach(word =>println(word._1 + ":" +word._2))
  }

}

File: bzz-spark/src/main/java/com/spark/study/SortWord.java
package com.spark.study;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.*;
import scala.Tuple2;

import java.util.Arrays;
import java.util.Iterator;

/**
 * @Title: SortWord
 * @ProjectName bzz-cloud
 * @Auther: clouder
 * @Date: 2018-07-13-23:40
 * @Description:
 */
public class SortWord {
	
	public static void main(String[] args) {
		SparkConf conf = new SparkConf().setMaster("local").setAppName("SortWord");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		JavaRDD<String> wordJavaRDD = jsc.textFile("D:\\app\\word.txt");
		
		JavaRDD<String> flatJavaRDD = wordJavaRDD.flatMap(new FlatMapFunction<String, String>() {
			@Override
			public Iterator<String> call(String s) throws Exception {
				String[] s1 = s.split(" ");
				return Arrays.asList(s1).iterator();
			}
		});
		
		JavaPairRDD<String, Integer> wordCountJavaPairRDD = flatJavaRDD.mapToPair(new PairFunction<String, String, Integer>() {
			@Override
			public Tuple2<String, Integer> call(String s) throws Exception {
				return new Tuple2<String, Integer>(s,1);
			}
		});
		JavaPairRDD<String, Integer> stringJavaPairRDD = wordCountJavaPairRDD.reduceByKey(new Function2<Integer, Integer, Integer>() {
			@Override
			public Integer call(Integer v1, Integer v2) throws Exception {
				return v1+v2;
			}
		});
		stringJavaPairRDD.foreach(new VoidFunction<Tuple2<String, Integer>>() {
			@Override
			public void call(Tuple2<String, Integer> v) throws Exception {
				System.out.println(v._1+" : "+v._2);
			}
		});
		
		
		JavaPairRDD<Integer, String> integerStringJavaPairRDD = stringJavaPairRDD.mapToPair(new PairFunction<Tuple2<String, Integer>, Integer, String>() {
			@Override
			public Tuple2<Integer, String> call(Tuple2<String, Integer> v) throws Exception {
				return new Tuple2<Integer, String>(v._2,v._1);
			}
		});
		JavaPairRDD<Integer, String> integerStringJavaPairRDD1 = integerStringJavaPairRDD.sortByKey(false);
		JavaPairRDD<String, Integer> resultRdd = integerStringJavaPairRDD1.mapToPair(new PairFunction<Tuple2<Integer, String>, String, Integer>() {
			@Override
			public Tuple2<String, Integer> call(Tuple2<Integer, String> v) throws Exception {
				return new Tuple2<String, Integer>(v._2, v._1);
			}
		});
		resultRdd.foreach(new VoidFunction<Tuple2<String, Integer>>() {
			@Override
			public void call(Tuple2<String, Integer> v) throws Exception {
				System.out.println(v._1+":"+v._2);
			}
		});
		
	}
}
